File: BinarySearchOnAnswer/AllocateBooks/AllocateBooks.java

public class AllocateBooks {

    /**
     * Problem: Allocate books to m students such that the maximum pages
     * allocated is minimized.
     */
    /**
     * Helper Method: Checks if allocation is possible with given maxPages.
     */
    private static boolean canAllocate(int[] arr, int m, int maxPages) {
        int students = 1;
        int pages = 0;

        for (int book : arr) {
            if (pages + book > maxPages) {
                students++;
                pages = book;
            } else {
                pages += book;
            }
        }

        return students <= m;
    }

    /**
     * Approach 1: Brute Force
     *
     * Idea: - Answer lies between max(arr) and sum(arr) - Try every possible
     * maxPages and check feasibility
     *
     * Time Complexity: O((sum - max) * n) Space Complexity: O(1)
     *
     * Stable: Yes
     */
    public static int allocateBooksBrute(int[] arr, int m) {
        if (m > arr.length) {
            return -1;
        }

        int max = 0;
        int sum = 0;

        for (int pages : arr) {
            max = Math.max(max, pages);
            sum += pages;
        }

        for (int maxPages = max; maxPages <= sum; maxPages++) {
            if (canAllocate(arr, m, maxPages)) {
                return maxPages;
            }
        }

        return -1;
    }

    /**
     * Approach 2: Binary Search on Answer (Optimal)
     *
     * Idea: - Minimum pages = max(arr) - Maximum pages = sum(arr) - If
     * allocation possible, try smaller maxPages
     *
     * Time Complexity: O(n * log(sum)) Space Complexity: O(1)
     *
     * Stable: Yes
     */
    public static int allocateBooksOptimal(int[] arr, int m) {
        if (m > arr.length) {
            return -1;
        }

        int low = 0;
        int high = 0;

        for (int pages : arr) {
            low = Math.max(low, pages);
            high += pages;
        }

        int ans = high;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (canAllocate(arr, m, mid)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return ans;
    }

    public static void main(String[] args) {
        int[] arr = {12, 34, 67, 90};
        int m = 2;

        System.out.println(allocateBooksBrute(arr, m));
        System.out.println(allocateBooksOptimal(arr, m));
    }
}

File: BinarySearchOnAnswer/AllocateBooks/AllocateBooks.ts
/**
 * Problem:
 * Allocate books to m students such that the
 * maximum pages assigned is minimized.
 */

class AllocateBooks {
  /**
   * Helper Method:
   * Checks feasibility for given maxPages
   */
  private static canAllocate(
    arr: number[],
    m: number,
    maxPages: number
  ): boolean {
    let students = 1;
    let pages = 0;

    for (const book of arr) {
      if (pages + book > maxPages) {
        students++;
        pages = book;
      } else {
        pages += book;
      }
    }

    return students <= m;
  }

  /**
   * Approach 1: Brute Force
   *
   * Time Complexity: O((sum - max) * n)
   * Space Complexity: O(1)
   *
   * Stable: Yes
   */
  static allocateBooksBrute(arr: number[], m: number): number {
    if (m > arr.length) return -1;

    let max = 0;
    let sum = 0;

    for (const pages of arr) {
      max = Math.max(max, pages);
      sum += pages;
    }

    for (let maxPages = max; maxPages <= sum; maxPages++) {
      if (this.canAllocate(arr, m, maxPages)) {
        return maxPages;
      }
    }

    return -1;
  }

  /**
   * Approach 2: Binary Search on Answer (Optimal)
   *
   * Time Complexity: O(n * log(sum))
   * Space Complexity: O(1)
   *
   * Stable: Yes
   */
  static allocateBooksOptimal(arr: number[], m: number): number {
    if (m > arr.length) return -1;

    let low = 0;
    let high = 0;

    for (const pages of arr) {
      low = Math.max(low, pages);
      high += pages;
    }

    let ans = high;

    while (low <= high) {
      const mid = Math.floor((low + high) / 2);

      if (this.canAllocate(arr, m, mid)) {
        ans = mid;
        high = mid - 1;
      } else {
        low = mid + 1;
      }
    }

    return ans;
  }
}

// Test
const arr = [12, 34, 67, 90];
const m = 2;

console.log(AllocateBooks.allocateBooksBrute(arr, m));
console.log(AllocateBooks.allocateBooksOptimal(arr, m));
